File: ios/Sources/CallKitPlugin/CallEventBus.swift
import Foundation

/// Persistent storage for the pending event queue (UserDefaults in production).
public protocol CallEventStore: AnyObject {
    func read() -> String?
    func write(_ json: String)
}

/// Queues call events until the plugin (and so the JS layer) is ready. Mirrors Android's
/// `CallEventBus`.
///
/// A VoIP push can wake a killed app and CallKit can answer the call before the WebView has
/// loaded. Events raised then are queued in order, persisted, and flushed when `CallKitPlugin`
/// attaches.
/// - Events restored in a new process are not replayed (the system ended those calls when the
///   old process died); they are exposed through `initialEvents()` with `restored: true`.
/// - Every event that happened before JS listened for it is also kept for `initialEvents()`
///   until `clearInitialEvents()` (callkeep's getInitialEvents / clearInitialEvents).
public final class CallEventBus {
    public struct Event {
        public let name: String
        public let data: [String: Any]
        /// Epoch milliseconds.
        public let timestamp: Double
        public let restored: Bool

        public func toDictionary() -> [String: Any] {
            return ["name": name, "data": data, "timestamp": timestamp, "restored": restored]
        }
    }

    /// Returns true when JS already had a listener for the event.
    public typealias Listener = (_ name: String, _ data: [String: Any]) -> Bool

    public static let shared = CallEventBus()

    static let maxPending = 50
    static let restoredMaxAgeMs: Double = 5 * 60 * 1000

    private let lock = NSLock()
    private var pending: [Event] = []
    private var initial: [Event] = []
    private var listener: Listener?
    private var owner: ObjectIdentifier?
    private var store: CallEventStore?
    var clock: () -> Double

    init(clock: @escaping () -> Double = { Date().timeIntervalSince1970 * 1000 }) {
        self.clock = clock
    }

    /// Install the persistent store. The first store installed restores what a previous
    /// process queued but never delivered.
    public func setStore(_ newStore: CallEventStore) {
        lock.lock()
        defer { lock.unlock() }
        let firstStore = store == nil
        store = newStore
        if firstStore {
            restore()
        }
    }

    public func emit(_ name: String, _ data: [String: Any] = [:]) {
        lock.lock()
        guard let target = listener else {
            if pending.count >= CallEventBus.maxPending {
                pending.removeFirst()
            }
            pending.append(Event(name: name, data: data, timestamp: clock(), restored: false))
            persist()
            lock.unlock()
            return
        }
        lock.unlock()

        if !target(name, data) {
            lock.lock()
            initial.append(Event(name: name, data: data, timestamp: clock(), restored: false))
            lock.unlock()
        }
    }

    /// Attach the listener and deliver everything queued while nobody was listening.
    public func attach(owner newOwner: AnyObject, _ newListener: @escaping Listener) {
        lock.lock()
        listener = newListener
        owner = ObjectIdentifier(newOwner)
        let toFlush = pending
        initial.append(contentsOf: toFlush)
        pending.removeAll()
        persist()
        lock.unlock()

        for event in toFlush {
            _ = newListener(event.name, event.data)
        }
    }

    public func detach(owner oldOwner: AnyObject) {
        lock.lock()
        defer { lock.unlock() }
        if owner == ObjectIdentifier(oldOwner) {
            listener = nil
            owner = nil
        }
    }

    public func initialEvents() -> [Event] {
        lock.lock()
        defer { lock.unlock() }
        return initial
    }

    public func clearInitialEvents() {
        lock.lock()
        defer { lock.unlock() }
        initial.removeAll()
    }

    func pendingSnapshot() -> [Event] {
        lock.lock()
        defer { lock.unlock() }
        return pending
    }

    // MARK: - Persistence (call with lock held)

    private func restore() {
        guard let json = store?.read(), let raw = json.data(using: .utf8) else {
            return
        }
        let now = clock()
        if let items = (try? JSONSerialization.jsonObject(with: raw)) as? [[String: Any]] {
            for item in items {
                guard let name = item["name"] as? String, let timestamp = (item["timestamp"] as? NSNumber)?.doubleValue else {
                    continue
                }
                if now - timestamp > CallEventBus.restoredMaxAgeMs {
                    continue
                }
                let data = item["data"] as? [String: Any] ?? [:]
                initial.append(Event(name: name, data: data, timestamp: timestamp, restored: true))
            }
        }
        store?.write("[]")
    }

    private func persist() {
        guard let store = store else {
            return
        }
        let items: [[String: Any]] = pending.compactMap { event in
            let item: [String: Any] = ["name": event.name, "data": event.data, "timestamp": event.timestamp]
            return JSONSerialization.isValidJSONObject(item) ? item : nil
        }
        if let raw = try? JSONSerialization.data(withJSONObject: items), let json = String(data: raw, encoding: .utf8) {
            store.write(json)
        }
    }
}

/// UserDefaults-backed store used by the plugin.
final class UserDefaultsEventStore: CallEventStore {
    private let key = "IonicCallkitPendingEvents"

    func read() -> String? {
        return UserDefaults.standard.string(forKey: key)
    }

    func write(_ json: String) {
        UserDefaults.standard.set(json, forKey: key)
    }
}

File: android/src/main/java/io/kreador/callkit/CallEventBus.java
package io.kreador.callkit;

import com.getcapacitor.JSObject;
import java.util.ArrayList;
import java.util.List;

/**
 * Replaces callkeep's RCTDeviceEventEmitter + delayedEvents queue.
 *
 * Native code (ConnectionService, FCM service, notification actions) can run
 * while no WebView / plugin instance exists, e.g. when a push wakes a killed app.
 * Events emitted then are queued and flushed once CallKitPlugin attaches.
 */
public final class CallEventBus {

    public interface Listener {
        void onCallEvent(String eventName, JSObject data);
    }

    public static final class Event {

        public final String name;
        public final JSObject data;

        Event(String name, JSObject data) {
            this.name = name;
            this.data = data;
        }
    }

    // Bound the queue so a device that never opens the app can't grow it forever.
    static final int MAX_PENDING = 50;

    private static final List<Event> pending = new ArrayList<>();
    private static Listener listener;

    private CallEventBus() {}

    public static void emit(String eventName, JSObject data) {
        Listener target;
        synchronized (CallEventBus.class) {
            target = listener;
            if (target == null) {
                if (pending.size() >= MAX_PENDING) {
                    pending.remove(0);
                }
                pending.add(new Event(eventName, data != null ? data : new JSObject()));
                return;
            }
        }
        target.onCallEvent(eventName, data != null ? data : new JSObject());
    }

    /** Attach the listener and deliver everything queued while nobody was listening. */
    public static void attach(Listener newListener) {
        List<Event> toFlush;
        synchronized (CallEventBus.class) {
            listener = newListener;
            toFlush = new ArrayList<>(pending);
            pending.clear();
        }
        for (Event event : toFlush) {
            newListener.onCallEvent(event.name, event.data);
        }
    }

    public static synchronized void detach(Listener oldListener) {
        if (listener == oldListener) {
            listener = null;
        }
    }

    static synchronized List<Event> pendingSnapshot() {
        return new ArrayList<>(pending);
    }

    static synchronized void reset() {
        pending.clear();
        listener = null;
    }
}
